### PR TITLE
WIP: Elaborate contribution guide

### LINKED ISSUE #82: Complete contribution guide

### PR DESCRIPTION
**Pull request by elevran**
The PR proposes a PR process for ClusterLink contributions.
It is meant (initially) for discussions.
The current version is incomplete (eg. marked with TODO where missing link/docs exist, there could be missing sections, etc.)
Depending on scope, we would expand the document in this PR or in follow up PRs.

Once agreed and merged this PR fixes #82

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -61,52 +68,75 @@ feedback on others’ ideas, and even sharing your own ideas, and experiences.
...
 
-<!-- ⚠️ **Explain how to set up a development environment** -->
+We have a [tutorial](TODO missing link) that walks you through setting up your developer
+ environment, making a change and testing it.
 
-## Sign Your Commits
+Here are the key steps, if you run into trouble, the tutorial has more details:
 
-[Instructions](https://contribute.cncf.io/maintainers/github/templates/required/contributing/#sign-your-commits)
+1. Install Go version 1.17 or higher.
```
[the remark this answers is not in the document]
Comment (elevran) [reply CONTRIBUTING.md:130]: These are minimal requirements. The CICD can run with a preferred Go version.
Do we know of limitations compiling with Go version <1.20?

[the remark this answers is not in the document]
Comment (orozery) [reply CONTRIBUTING.md:130]: > These are minimal requirements. The CICD can run with a preferred Go version. Do we know of limitations compiling with Go version <1.20?

Yep

Comment (elevran) [inline CONTRIBUTING.md:130]:
```suggestion
1. Install Go version 1.20 or higher.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -10,22 +9,30 @@
...
-bug report and let us know!
+The ClusterLink community is governed by our [Code of Conduct](./CODE_OF_CONDUCT.md)
+ and we expect all contributors to abide by it.
 
 ## Ways to Contribute
 
-[Instructions](https://contribute.cncf.io/maintainers/github/templates/required/contributing/#ways-to-contribute)
+There are more ways to help than code contribution. If you have experience in marketing,
+ content creation, technical writing, project management, community management, or other
+ areas we might not be have considered - please reach out!
```
Comment (zivnevo) [inline CONTRIBUTING.md:35]:
```suggestion
 areas we might not have considered - please reach out!
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -61,52 +68,75 @@ feedback on others’ ideas, and even sharing your own ideas, and experiences.
 
 ## Find an Issue
 
-[Instructions](https://contribute.cncf.io/maintainers/github/templates/required/contributing/#find-an-issue)
+If you're new to ClusterLink, you are in the best position to give us feedback on areas of
+our project that we need help with including:
```
Comment (zivnevo) [inline CONTRIBUTING.md:72]:
```suggestion
our project that we can improve, including:
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -61,52 +68,75 @@ feedback on others’ ideas, and even sharing your own ideas, and experiences.
...
+* Bugs in our automation scripts
+
+If anything doesn't make sense, or doesn't work when you run it, please open a
+bug report and let us know!
 
 We have good first issues for new contributors and help wanted issues suitable
-for any contributor. [good first issue](TODO) has extra information to
-help you make your first contribution. [help wanted](TODO) are issues
+for any contributor. [good first issue](TODO missing link) has extra information to
+help you make your first contribution. [help wanted](TODO missing link) are issues
```
Comment (zivnevo) [inline CONTRIBUTING.md:83]:
```suggestion
help you make your first contribution. [help wanted](https://github.com/clusterlink-net/clusterlink/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22) are issues
```
Why not this link?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -61,52 +68,75 @@ feedback on others’ ideas, and even sharing your own ideas, and experiences.
...
+We have a [roadmap](TODO missing link) that will give you a good idea of the larger
+ features that we are working on right now. That may help you decide what you would
+ like to work on after you have tackled an issue or two. If you have a big idea for
+ ClusterLink, you can propose it by creating an issue and marking it `enhancement`.
 
-The best way to reach us with a question when contributing is to ask on:
+## Asking for Help
 
-<!-- ⚠️ **Pick the way(s) that you prefer people ask for help** -->
+All contributors will might get stuck sometimes. The best way to reach us with a question
```
Comment (zivnevo) [inline CONTRIBUTING.md:107]:
```suggestion
All contributors might get stuck sometimes. The best way to reach us with a question
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -115,28 +145,207 @@ our project's repositories. The [Developer Certificate of Origin
...
-## Pull Request Checklist
+## Pull Request (PR) Lifecycle
+
+### Scope of a PR
+
+<!-- 
+What kind of pull requests do you prefer: small scope, incremental value or feature complete?
+-->
+When you are ready to start on a unit of work, such as fixing a bug or implementing a
+ feature, create a branch. Each branch (and ultimately each PR) should represent a
```
Comment (zivnevo) [inline CONTRIBUTING.md:182]:
```suggestion
 feature, create a branch on your fork. Each branch (and ultimately each PR) should represent a
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -115,28 +145,207 @@ our project's repositories. The [Developer Certificate of Origin
...
+ refactoring - please do these on different branches and in different PRs. While this
+ is a bit of a hassle for you, it makes your changes much easier to review and faster
+ to accept.
+ Additionally, a change history with focused commits is easier to work with in the future,
+ test failures are easier to localize and interaction between changes is easier to see.
+
+### When to Open a PR
+
+While it's OK to submit a PR directly for problems such as typos or other things where
+the motivation/problem is unambiguous, most PRs should have an associate github issue.
```
Comment (zivnevo) [inline CONTRIBUTING.md:193]:
```suggestion
the motivation/problem is unambiguous, most PRs should have an associated GitHub Issue.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -115,28 +145,207 @@ our project's repositories. The [Developer Certificate of Origin
...
+git checkout -b MY_FEATURE_BRANCH main
+```
+
+### The Life of Pi (oops! The Life of a PR)
+
+1. Start on a [unit of work](#scope-of-a-pr) by updating from upstream and then creating
+ a branch off of it. Please use a descriptive name (e.g., `fix-issue-17` or `lb-policy-support`). Please do **not** create dependent branches and always branch from
+ an up-to-date `upstream/main`.
+
+1. You create a draft or WIP pull request at any time. Reviewers will ignore it mostly
```
Comment (zivnevo) [inline CONTRIBUTING.md:222]:
```suggestion
1. Create a draft or WIP pull request at any time. Reviewers will ignore it mostly
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -115,28 +145,207 @@ our project's repositories. The [Developer Certificate of Origin
...
+ unless you mention someone and ask for help. Feel free to open one and use the pull
+ request to see if the CI passes. Once you are ready for a review, remove the WIP or
+ click "Ready for Review" and leave a comment that it's ready for review. You may also
+ mark the PR as not ready by assigning the label `do-not-merge/wip` to it.
+ Give the PR a descriptive title, that would be appropriate as a commit message once
+ your work when merged. Include additional information, as appropriate, in the PR
```
Comment (zivnevo) [inline CONTRIBUTING.md:228]:
```suggestion
 your work is merged. Include additional information, as appropriate, in the PR
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -115,28 +145,207 @@ our project's repositories. The [Developer Certificate of Origin
...
+
+1. Start on a [unit of work](#scope-of-a-pr) by updating from upstream and then creating
+ a branch off of it. Please use a descriptive name (e.g., `fix-issue-17` or `lb-policy-support`). Please do **not** create dependent branches and always branch from
+ an up-to-date `upstream/main`.
+
+1. You create a draft or WIP pull request at any time. Reviewers will ignore it mostly
+ unless you mention someone and ask for help. Feel free to open one and use the pull
+ request to see if the CI passes. Once you are ready for a review, remove the WIP or
+ click "Ready for Review" and leave a comment that it's ready for review. You may also
+ mark the PR as not ready by assigning the label `do-not-merge/wip` to it.
```
Comment (zivnevo) [inline CONTRIBUTING.md:226]: Move this sentence to the beginning of the paragraph

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -115,28 +145,207 @@ our project's repositories. The [Developer Certificate of Origin
...
+ by clarifying the code or add documentation, rather than in the PR comment thread.
+
+1. A reviewer will assign themselves to the pull request. If you don't see anyone
+ assigned after 3 business days, you can leave a comment asking for a review, or ping in
+ slack. Sometimes we have busy days, sick days, weekends and vacations, so a little
+ patience is appreciated! 🙇‍♀️
+
+1. The reviewer will leave feedback.
+
+* `nits``: These are suggestions that you may decide to incorporate into your pull
```
Comment (zivnevo) [inline CONTRIBUTING.md:246]:
```suggestion
* ``nits``: These are suggestions that you may decide to incorporate into your pull
```
?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+* Write a comment in the conversation in the GitHub pull request.
+
+1. When a pull request has been approved, the reviewer will squash and merge your
+ commits. If you prefer to rebase your own commits, at any time leave a comment on the
+ pull request to let them know that.
+
+1. At this point your changes are available to be included in the next release of
+ ClusterLink! After your first pull request is merged, you will be invited to the
+ Contributors team (TODO create Github team) which you may choose to accept (or not).
+ Joining the team lets you have issues in GitHub assigned to you.
```
Comment (zivnevo) [inline CONTRIBUTING.md:300]: This is a bit odd - numbered bullets coming right after plain bullets.

**Review by zivnevo — COMMENTED** [review]
Some more tiny things

File: CONTRIBUTING.md (hunk 12 above)
Comment (elevran) [reply CONTRIBUTING.md:300]: It was meant as continuation of the previous steps. Changed to use headings instead

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+ you can associate it using "linked issues". Either of these links your PR to the issue
+ and automatically closes the issue when the PR is accepted.
+ Do **not** use the PR comments to document the code as they will most likely never be
+ seen after the PR is merged. Please **do** make your code self-explanatory. If there is
+ information that is needed by a programmer reading the source code, you should put it in
+ a code comment. This also applies to answering questions from reviewers: it is best done
+ by clarifying the code or add documentation, rather than in the PR comment thread.
+
+1. A reviewer will assign themselves to the pull request. If you don't see anyone
+ assigned after 3 business days, you can leave a comment asking for a review, or ping in
```
Comment (orozery) [inline CONTRIBUTING.md:246]: Feels too formal to me :)

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+ your work is merged. Include additional information, as appropriate, in the PR
+ description paragraph to explain what you have done.
+ If your PR fixes any issues, the description should refer to it (e.g., "fixes #NNN") or
+ you can associate it using "linked issues". Either of these links your PR to the issue
+ and automatically closes the issue when the PR is accepted.
+ Do **not** use the PR comments to document the code as they will most likely never be
+ seen after the PR is merged. Please **do** make your code self-explanatory. If there is
+ information that is needed by a programmer reading the source code, you should put it in
+ a code comment. This also applies to answering questions from reviewers: it is best done
+ by clarifying the code or add documentation, rather than in the PR comment thread.
```
Comment (orozery) [inline CONTRIBUTING.md:241]: This paragraph is too long.
I would remove the last part (starting from `Do not`).

Code readability should be explained along with style, testing, etc somewhere else.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+ GitHub code review system. Simply resolve the conversation. Respond if more discussion
+ is needed, such as asking a follow-up question or explaining why you disagree with the
+ suggestion. If the reviewer is asking a question, then usually the best way to answer
+ it is by improving the code or documentation. Answering it only in the code review
+ will not help future programmers after the PR is merged.
+
+Respond to the feedback by making changes in your working copy, committing them, and
+ periodically pushing them to GitHub when the tests pass locally. As soon as you
+ receive feedback, you can start working on it. The reviewer should assign the code
+ review back to you, but they might forget, so don't wait for that.
```
Comment (orozery) [inline CONTRIBUTING.md:268]: I don't think this is necessary (`The reviewer should`...).

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+
+After you have made the changes (in new commits please!), leave a comment asking the
+ the reviewer to take another look. If 3 business days go by with no review, it is okay
```
Comment (orozery) [inline CONTRIBUTING.md:271]: ditto, too formal :)

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+ to bump. With the exception of rebasing discussed below, please do **not** force a push
+ with `git push -f` as it might cause loss of code review comments and context associated
+ with previous commits.
+
+When the PR is ready for merging, you may need to rebase your PR on top of the current codebase.
+This can be done via the PR's GitHub Web page or with the CLI as either `git pull --rebase`
+or `git fetch` and `git rebase` from `upstream/main`.
+If there are any conflicts please fix them locally before committing and pushing to the PR branch.
+If you're unsure on how to resolve a conflict, please ask the reviewer for guidance or help.
```
Comment (orozery) [inline CONTRIBUTING.md:280]: I would remove this.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -115,28 +144,231 @@ our project's repositories. The [Developer Certificate of Origin
...
+The repository owner can prevent incorrect pull request merges. In the repository settings, in the “Merge button” section, disable “Allow merge commits” and “Allow rebase merging”. You might also want to enable “Automatically delete head branches”.
+-->
+
+After you have addressed all the review feedback, explicitly request a re-review.
+ Do not assume that person will know when you are done. There are many ways to request
+ a re-review:
+
+* Add the reviewer to the PR on Github - this works even if that person has reviewed the
+ pull request before.
+* Assign the pull request to that person, using the “Assignees” list.
```
Comment (orozery) [inline CONTRIBUTING.md:296]: Don't really like this :)

File: CONTRIBUTING.md (hunk 12 above)
Comment (orozery) [inline CONTRIBUTING.md:310]: I would remove this.

**Comment by orozery** [comment]
Forgot to add:
My "nit: trivial" comments are for things I feel make this document too-much user-friendly.

File: CONTRIBUTING.md (hunk 17 above)
Comment (elevran) [reply CONTRIBUTING.md:280]: clarification: remove just the last sentence, regarding resolving the conflict?

File: CONTRIBUTING.md (hunk 12 above)
Comment (elevran) [reply CONTRIBUTING.md:310]: This is meant as a way to recognize contributor efforts. It does not grant maintainer rights.
AFAIK, you can't assign issues (or mark them as assigned to yourself) if you're not somehow part of the repo's contributors list

**Comment by elevran** [comment]
@orozery I'm trying to find the right level of instructions between TMI and not enough information for new contributors.
I think the nits take the "less is more". I'll take a look at the level of detail in other projects (e.g., k8s, Istio, ...) for reference and then keep/remove the "trivial" steps.

File: CONTRIBUTING.md (hunk 16 above)
Comment (elevran) [reply CONTRIBUTING.md:271]: As a contributor I would like to know if it is ok to nudge the maintainers or not and if so, when?
But maybe that's just me ;-)

File: CONTRIBUTING.md (hunk 17 above)
Comment (orozery) [reply CONTRIBUTING.md:279]: The last sentence I would remove.
As for the block, I would remove it as I think it is trivial.

File: CONTRIBUTING.md (hunk 12 above)
Comment (orozery) [reply CONTRIBUTING.md:309]: I don't think you want to ALWAYS want to add anyone who upstreamed a commit.
Perhaps they're not even interested in that honor.
And on the other direction, the team may not want to necessarily add anyone who upstreamed a change.
This also relates to my "nit: trivial" comments.
I don't think it makes a good impression if you say "we accept anyone who makes a single commit".

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -61,52 +68,74 @@ feedback on others’ ideas, and even sharing your own ideas, and experiences.
...
+on.  
 Sometimes there won’t be any issues with these labels. That’s ok! There is
 likely still something for you to work on. If you want to contribute but you
-don’t know where to start or can't find a suitable issue, you can ⚠️ **explain how people can ask for an issue to work on**.
+don’t know where to start or can't find a suitable issue, you can ⚠️ **explain how people 
+can ask for an issue to work on**.
 -->
 
 Once you see an issue that you'd like to work on, please post a comment saying
 that you want to work on it. Something like "I want to work on this" is fine.
```
Comment (michalmalka) [inline CONTRIBUTING.md:97]: Can remove the "something like...." it is obvious...

**Comment by elevran** [comment]
Having looked at multiple CONTRIBUTING (k8s, Istio, Envoy, Linkerd), there seems to be a variety, with different levels of details.
Most maintain a relatively terse CONTRIBUTING guide but do provide considerably more detail in related documents (e.g., from k8s - there is a [high level contributing guide]https://github.com/kubernetes/community/tree/master/contributors/guide) as well as [very detailed PR workflow](https://github.com/kubernetes/community/blob/master/contributors/guide/github-workflow.md).
Envoy is another [example](https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md).

I will close this PR in favor of multiple documents with `CONTRIBUTING.md` serving as more of a ToC for the detailed documents that will follow.

File: CONTRIBUTING.md (hunk 12 above)
Comment (elevran) [reply CONTRIBUTING.md:309]: Some projects do that, others do not. Some even extend an offer of maintainership to committers. Submitters are free to refuse...
I'm not proposing maintainers, only contributor. They can also decide to join the developer mailing list...

Assuming you want to acknowledge contributions publicly, you would typically do this in
- contributors file, listing everyone
- listing relevant contributions in release notes
- adding them to contributors mailing list.

We can change it to >1 commit, but if someone wants to help out with the project on a more ongoing basis, you may want to assign issues to them at some point.

**Pull request closed without merging**